Lib/compression/zstd/__init__.py

```diff
@@ -0,0 +1,293 @@
+    optional arguments *zstd_dict*, *options*.
+
+    For incremental decompression, use an ZstdDecompressor instead.
+    """
+    results = []
+    while True:
+        decomp = ZstdDecompressor(options=options, zstd_dict=zstd_dict)
+        try:
+            res = decomp.decompress(data)
+        except ZstdError:
+            if results:
+                break  # Leftover data is not a valid LZMA/XZ stream; ignore it.
+            else:
+                raise  # Error on the first iteration; bail out.
+        results.append(res)
+        if not decomp.eof:
+            raise ZstdError("Compressed data ended before the "
+                            "end-of-stream marker was reached")
+        data = decomp.unused_data
+        if not data:
+            break
+    return b"".join(results)
+
+class _UnsupportedCParameter:
+    def __set_name__(self, _, name):
+        self.name = name
+
+    def __get__(self, *_, **__):
+        msg = ("%s CParameter not available, zstd version is %s.") % (
+            self.name,
+            zstd_version,
+        )
+        raise NotImplementedError(msg)
+
+
+class CParameter(enum.IntEnum):
+    """Compression parameters"""
+
+    compressionLevel = _zstd._ZSTD_c_compressionLevel
+    windowLog = _zstd._ZSTD_c_windowLog
+    hashLog = _zstd._ZSTD_c_hashLog
+    chainLog = _zstd._ZSTD_c_chainLog
+    searchLog = _zstd._ZSTD_c_searchLog
+    minMatch = _zstd._ZSTD_c_minMatch
+    targetLength = _zstd._ZSTD_c_targetLength
+    strategy = _zstd._ZSTD_c_strategy
+
+    targetCBlockSize = _UnsupportedCParameter()
+
+    enableLongDistanceMatching = _zstd._ZSTD_c_enableLongDistanceMatching
+    ldmHashLog = _zstd._ZSTD_c_ldmHashLog
+    ldmMinMatch = _zstd._ZSTD_c_ldmMinMatch
+    ldmBucketSizeLog = _zstd._ZSTD_c_ldmBucketSizeLog
+    ldmHashRateLog = _zstd._ZSTD_c_ldmHashRateLog
+
+    contentSizeFlag = _zstd._ZSTD_c_contentSizeFlag
+    checksumFlag = _zstd._ZSTD_c_checksumFlag
+    dictIDFlag = _zstd._ZSTD_c_dictIDFlag
+
+    nbWorkers = _zstd._ZSTD_c_nbWorkers
+    jobSize = _zstd._ZSTD_c_jobSize
+    overlapLog = _zstd._ZSTD_c_overlapLog
+
+    @functools.lru_cache(maxsize=None)
+    def bounds(self):
+        """Return lower and upper bounds of a compression parameter, both inclusive."""
+        # 1 means compression parameter
+        return _zstd._get_param_bounds(1, self.value)
+
+
+class DParameter(enum.IntEnum):
+    """Decompression parameters"""
+
+    windowLogMax = _zstd._ZSTD_d_windowLogMax
+
+    @functools.lru_cache(maxsize=None)
+    def bounds(self):
+        """Return lower and upper bounds of a decompression parameter, both inclusive."""
+        # 0 means decompression parameter
+        return _zstd._get_param_bounds(0, self.value)
+
+
+class Strategy(enum.IntEnum):
+    """Compression strategies, listed from fastest to strongest.
+
+    Note : new strategies _might_ be added in the future, only the order
+    (from fast to strong) is guaranteed.
```

Comment:
This makes me feel a little uneasy, if we expose a numeric value (`Strategy` is an IntEnum) people will end up depending on it.
```suggestion
    Note that new strategies might be added in the future.
    Only the order (from fast to strong) is guaranteed,
    the numeric value might change.
```

Reply:
I don't love this either, but this comes from libzstd itself:

> /* note : new strategies _might_ be added in the future.
>                         Only the order (from fast to strong) is guaranteed */

https://facebook.github.io/zstd/doc/api_manual_v1.5.7.html

I'm not sure if there is a way we could expose this in a stable manner on top of an unstable interface from libzstd. I think the best we can do is document it loudly.

Reply:
Even with plain numbers, the same has always been true in all compression libraries.  it is up to the underlying implementation as to what they mean, and some do change their meaning at times.  the ordering is indeed the only thing they all seem to retain.  ex: zlib and zlib-ng differ on some meanings as zlib-ng decided to try to match an un-named user intent on some numbers.  I like that zstd is at least giving names to some values.  that feels more explicit when read in code.  what happens under the hood is up to the library regardless.